File: 백준/Silver/1012. 유기농 배추/유기농 배추.cc
#include <bits/stdc++.h>
using namespace std;

const int dy[] = {-1, 0, 1, 0};
const int dx[] = {0, 1, 0, -1};
int t, m, n, k, y, x, ret, a[54][54], visited[54][54];

void dfs(int y, int x) {
  visited[y][x] = 1;

  for (int i = 0; i < 4; i++) {
    int ny = y + dy[i];
    int nx = x + dx[i];

    if (ny < 0 || nx < 0 || ny > n || nx > m) continue;
    if (a[ny][nx] == 0 || visited[ny][nx]) continue;
    dfs(ny, nx);
  }
}

int main() {
  cin >> t;

  for (int i = 0; i < t; i++) {
    memset(visited, 0, sizeof(visited));
    memset(a, 0, sizeof(a));
    ret = 0;

    cin >> m >> n >> k;
    for (int j = 0; j < k; j++) {
      cin >> x >> y;
      a[y][x] = 1;
    }
    for (int j = 0; j < n; j++) {
      for (int l = 0; l < m; l++) {
        if (a[j][l] && visited[j][l] == 0) {
          dfs(j, l);
          ret++;
        }
      }
    }
    cout << ret << '\n';
  }

  return 0;
}

File: 백준/Silver/1012. 유기농 배추/유기농 배추.js
const fs = require('fs');
// const input = fs.readFileSync('./input.txt').toString().trim().split('\n');
const input = fs.readFileSync('/dev/stdin').toString().trim().split('\n');

const t = Number(input[0])
let line = 1;

const dy = [-1, 0, 1, 0];
const dx = [0, 1, 0, -1];

const dfs = (y, x, visited, n, m, a) =>{
    visited[y][x] = 1;
    
    for(let i=0; i<4; i++){
        const ny = y + dy[i];
        const nx = x + dx[i];

        if(ny <0 || nx < 0|| ny >= n || nx >= m || visited[ny][nx] || !a[ny][nx]) continue;
        dfs(ny, nx, visited, n, m, a);
    }
}

for (let i=0; i<t; i++){
    const [n, m, k] = input[line++].split(' ').map(Number);

    let ret = 0;
    const a = Array.from({length: n}, () => Array(m).fill(0));
    const visited = Array.from({length: n}, () => Array(m).fill(0));

    for (let j=0; j<k; j++){
        const [y, x] = input[line++].split(' ').map(Number);
        a[y][x] = 1;
    }

    // 로직
    for(let i=0; i<n; i++){
        for(let j=0; j<m; j++){
            if(visited[i][j] || a[i][j] === 0) continue;
            dfs(i, j, visited, n, m, a);
            ret++;
        }
    }

    console.log(ret);
}